dashboard/app/reporting_email.go

```diff
@@ -459,7 +467,111 @@ func loadBugInfo(c context.Context, msg *email.Email) *bugInfoResult {
+	title, seq, err := subjectParser.parseTitle(subject)
+	if err != nil {
+		return nil, err
+	}
+	// Query all bugs with this title.
+	var bugs []*Bug
+	bugKeys, err := db.NewQuery("Bug").
+		Filter("Title=", title).
+		GetAll(c, &bugs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch bugs: %v", err)
+	}
+	// Filter the bugs by the email.
+	candidates := []*bugInfoResult{}
+	for i, bug := range bugs {
+		log.Infof(c, "processing bug %v", bug.displayTitle())
+		// We could add it to the query, but it's probably not worth it - we already have
+		// tons of db indexes while the number of matching bugs should not be large anyway.
+		if bug.Seq != int64(seq) {
+			log.Infof(c, "bug's seq is %v, wanted %d", bug.Seq, seq)
+			continue
+		}
+		if bug.sanitizeAccess(AccessPublic) != AccessPublic {
+			log.Infof(c, "access denied")
+			continue
+		}
+		reporting, bugReporting, _, _, err := currentReporting(c, bug)
+		if err != nil || reporting == nil {
+			log.Infof(c, "could not query reporting: %s", err)
+			continue
+		}
+		emailConfig, ok := reporting.Config.(*EmailConfig)
+		if !ok {
+			log.Infof(c, "reporting is not EmailConfig (%q)", subject)
+			continue
+		}
+		if emailConfig.Email == listID {
+			candidates = append(candidates, &bugInfoResult{
+				bug: bug, bugKey: bugKeys[i],
+				bugReporting: bugReporting, reporting: reporting,
+			})
+		} else {
+			log.Infof(c, "config's Email is %v, wanted %v", emailConfig.Email, listID)
+		}
+	}
+	if len(candidates) == 1 {
+		return candidates[0], nil
+	}
+	return nil, fmt.Errorf("unable to determine the bug")
+}
+
+type subjectTitleParser struct {
+	withSeq      *regexp.Regexp
+	noSeq        *regexp.Regexp
+	regexpsReady sync.Once
+}
+
+func (p *subjectTitleParser) parseTitle(subject string) (string, int, error) {
+	p.prepareRegexps()
+	subject = strings.TrimSpace(subject)
+	parts := p.withSeq.FindStringSubmatch(subject)
+	if parts == nil {
+		parts = p.noSeq.FindStringSubmatch(subject)
 	}
+	if parts == nil || parts[1] == "" {
+		return "", 0, fmt.Errorf("failed to extract the title")
+	}
+	title := parts[1]
+	seq := 0
+	if len(parts) == 3 {
+		rawSeq, err := strconv.Atoi(parts[2])
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to parse seq: %w", err)
+		}
+		seq = rawSeq - 1
+	}
+	return title, seq, nil
+}
+
+func (p *subjectTitleParser) prepareRegexps() {
+	p.regexpsReady.Do(func() {
+		stripPrefixes := []string{`R[eE]:`}
+		for _, ns := range config.Namespaces {
+			for _, rep := range ns.Reporting {
+				emailConfig, ok := rep.Config.(*EmailConfig)
+				if !ok {
+					continue
+				}
+				if ok && emailConfig.SubjectPrefix != "" {
+					stripPrefixes = append(stripPrefixes,
+						regexp.QuoteMeta(emailConfig.SubjectPrefix))
+				}
+			}
+		}
+		rePrefixes := `^(?:(?:` + strings.Join(stripPrefixes, "|") + `)\s*)*`
+		p.noSeq = regexp.MustCompile(rePrefixes + `(.*)$`)
+		p.withSeq = regexp.MustCompile(rePrefixes + `(.*)\s\((\d+)\)$`)
```

Comment:
Both cases can be handled with a single regexp `(.*?)(?:\s\((\d+)\))?$`
where `*?` is a non-greedy match

Reply:
Interesting! I thought regexps in Go do not support anything non-greedy. Adjusted the code.